From 81a86844223e4f106370c90b8cc8e3d9b2f9b88d Mon Sep 17 00:00:00 2001
From: "Paul E. McKenney" <paulmck@linux.vnet.ibm.com>
Date: Wed, 1 Apr 2015 08:19:59 -0700
Subject: [PATCH] powerpc: Fix smp_mb__before_spinlock()

References: bsc#1110247
Patch-mainline: v4.2-rc1
Git-commit: a76ff6884bfedfafcbb0d9c84c7a6b6a546cba6d

Currently, smp_mb__before_spinlock() is defined to be smp_wmb()
in core code, but this is not sufficient on PowerPC.  This patch
therefore supplies an override for the generic definition to
strengthen smp_mb__before_spinlock() to smp_mb(), as is needed
on PowerPC.

Signed-off-by: Paul E. McKenney <paulmck@linux.vnet.ibm.com>
Cc: <linuxppc-dev@lists.ozlabs.org>
Acked-by: Michal Suchanek <msuchanek@suse.de>
---
 arch/powerpc/include/asm/atomic.h   | 1 +
 arch/powerpc/include/asm/spinlock.h | 1 +
 2 files changed, 2 insertions(+)

diff --git a/arch/powerpc/include/asm/atomic.h b/arch/powerpc/include/asm/atomic.h
index b06172e3210a..dbd2c6b7b8f7 100644
--- a/arch/powerpc/include/asm/atomic.h
+++ b/arch/powerpc/include/asm/atomic.h
@@ -249,6 +249,7 @@ static __inline__ int atomic_dec_if_positive(atomic_t *v)
 #define smp_mb__after_atomic_dec()      smp_mb()
 #define smp_mb__before_atomic_inc()     smp_mb()
 #define smp_mb__after_atomic_inc()      smp_mb()
+#define smp_mb__before_spinlock()   smp_mb()
 
 #ifdef __powerpc64__
 
diff --git a/arch/powerpc/include/asm/spinlock.h b/arch/powerpc/include/asm/spinlock.h
index 391c9ce79f4c..3dc75d985a5f 100644
--- a/arch/powerpc/include/asm/spinlock.h
+++ b/arch/powerpc/include/asm/spinlock.h
@@ -28,6 +28,7 @@
 #include <asm/asm-compat.h>
 #include <asm/synch.h>
 #include <asm/ppc-opcode.h>
+#include <asm/atomic.h> /* PPC specific smp_mb__before_spinlock */
 
 #ifdef CONFIG_PPC64
 /* use 0x800000yy when locked, where yy == CPU number */
-- 
2.13.7

